src/arch/riscv/kernel/vspace.c

```diff
@@ -71,52 +110,68 @@ map_kernel_window(void)
     /* mapping of kernelBase (virtual address) to kernel's physBase  */
     assert(CONFIG_PT_LEVELS > 1 && CONFIG_PT_LEVELS <= 4);
 
-    /* kernel window starts at PPTR_BASE */
-    word_t pptr = PPTR_BASE;
+    word_t pptr = PPTR_BASE;   /* kernel window starts at PPTR_BASE */
+    word_t paddr = PADDR_BASE; /* first we map in memory from PADDR_BASE */
+
+    word_t bits = sizeof(word_t) * CHAR_BITS;
+    word_t max_address = BIT(bits) - 1;
 
-    /* first we map in memory from PADDR_BASE */
-    word_t paddr = PADDR_BASE;
     while (pptr < KERNEL_BASE) {
         assert(IS_ALIGNED(pptr, RISCV_GET_LVL_PGSIZE_BITS(1)));
         assert(IS_ALIGNED(paddr, RISCV_GET_LVL_PGSIZE_BITS(1)));
-        kernel_pageTables[0][RISCV_GET_PT_INDEX(pptr, 1)] =
-            pte_new(
-                paddr >> seL4_PageBits,
-                0,  /* sw */
-                1,  /* dirty */
-                1,  /* accessed */
-                1,  /* global */
-                0,  /* user */
-                1,  /* execute */
-                1,  /* write */
-                1,  /* read */
-                1   /* valid */
-            );
-        pptr += RISCV_GET_LVL_PGSIZE(1);
-        paddr += RISCV_GET_LVL_PGSIZE(1);
+
+        kernel_Level1_pageTable[RISCV_GET_PT_INDEX(pptr, PT_LEVEL_1)] = pte_next(paddr, true);
+
+        pptr += RISCV_GET_LVL_PGSIZE(PT_LEVEL_1);
+        paddr += RISCV_GET_LVL_PGSIZE(PT_LEVEL_1);
     }
+
     /* now we should be mapping the kernel base, starting again from PADDR_LOAD */
     assert(pptr == KERNEL_BASE);
     paddr = PADDR_LOAD;
-    /* pptr will overflow at the end of the address range so this loop condition looks odd */
-    while (pptr >= KERNEL_BASE) {
-        assert(IS_ALIGNED(pptr, RISCV_GET_LVL_PGSIZE_BITS(1)));
-        assert(IS_ALIGNED(paddr, RISCV_GET_LVL_PGSIZE_BITS(1)));
-        kernel_pageTables[0][RISCV_GET_PT_INDEX(pptr, 1)] =
-            pte_new(
-                paddr >> seL4_PageBits,
-                0,  /* sw */
-                1,  /* dirty */
-                1,  /* accessed */
-                1,  /* global */
-                0,  /* user */
-                1,  /* execute */
-                1,  /* write */
-                1,  /* read */
-                1   /* valid */
-            );
-        pptr += RISCV_GET_LVL_PGSIZE(1);
-        paddr += RISCV_GET_LVL_PGSIZE(1);
+
+#if CONFIG_PT_LEVELS == 3
+    word_t orig_indx_1GiB = RISCV_GET_PT_INDEX(pptr, PT_LEVEL_1);
+#endif
+
+    while (pptr < max_address && pptr >= KERNEL_BASE) {
+
+        /* if both phy and virt address are aligned on 1GiB boundary, use Level 1 mappings */
```

Comment:
This comment only applies to Sv39

Reply:
I'll specify for Sv39, then add a line explaining that this checks for 2MiB alignment in Sv32.